library/runtime/src/commonMain/kotlin/io/matthewnelson/kmp/tor/runtime/internal/AbstractRuntimeEventProcessor.kt

```diff
@@ -0,0 +1,143 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+package io.matthewnelson.kmp.tor.runtime.internal
+
+import io.matthewnelson.kmp.tor.core.api.annotation.InternalKmpTorApi
+import io.matthewnelson.kmp.tor.core.resource.SynchronizedObject
+import io.matthewnelson.kmp.tor.core.resource.synchronized
+import io.matthewnelson.kmp.tor.runtime.RuntimeEvent
+import io.matthewnelson.kmp.tor.runtime.ctrl.AbstractTorEventProcessor
+import io.matthewnelson.kmp.tor.runtime.ctrl.api.ItBlock
+import io.matthewnelson.kmp.tor.runtime.ctrl.api.TorEvent
+
+@OptIn(InternalKmpTorApi::class)
+internal abstract class AbstractRuntimeEventProcessor(
+    staticTag: String?,
+    initialObservers: Set<RuntimeEvent.Observer<*>>,
+    initialTorEventObservers: Set<TorEvent.Observer>,
+):  AbstractTorEventProcessor(staticTag, initialTorEventObservers),
+    RuntimeEvent.Processor
+{
+
+    private val observers = initialObservers.toMutableSet()
+
+    private val lock = SynchronizedObject()
+
+    public final override fun add(observer: RuntimeEvent.Observer<*>) {
+        withRuntimeObservers { add(observer) }
+    }
+
+    public final override fun add(vararg observers: RuntimeEvent.Observer<*>) {
+        if (observers.isEmpty()) return
+        withRuntimeObservers { observers.forEach { add(it) } }
+    }
+
+    public final override fun remove(observer: RuntimeEvent.Observer<*>) {
+        withRuntimeObservers { remove(observer) }
+    }
+
+    public final override fun remove(vararg observers: RuntimeEvent.Observer<*>) {
+        if (observers.isEmpty()) return
+        withRuntimeObservers { observers.forEach { remove(it) } }
+    }
+
+    public final override fun removeAll(event: RuntimeEvent<*>) {
+        withRuntimeObservers {
+            val iterator = iterator()
+            while (iterator.hasNext()) {
+                val observer = iterator.next()
+                if (staticTag != null && observer.tag == staticTag) continue
+
+                if (observer.event == event) {
+                    iterator.remove()
+                }
+            }
+        }
+    }
+
+    public final override fun removeAll(vararg events: RuntimeEvent<*>) {
+        withRuntimeObservers {
+            val iterator = iterator()
+            while (iterator.hasNext()) {
+                val observer = iterator.next()
+                if (staticTag != null && observer.tag == staticTag) continue
+
+                if (events.contains(observer.event)) {
+                    iterator.remove()
+                }
+            }
+        }
+    }
+
+    public final override fun removeAll(tag: String) {
+        if (staticTag != null && tag == staticTag) return
+
+        withRuntimeObservers {
+            val iterator = iterator()
+            while (iterator.hasNext()) {
+                val observer = iterator.next()
+                if (staticTag != null && observer.tag == staticTag) continue
```

Comment:
Move to protected extension function in `AbstractTorEventProcessor`

```kotlin
protected fun String.isStaticTag(): Boolean = staticTag != null && this == staticTag
```

Usage
```kotlin
if (observer.tag.isStaticTag()) continut
```